bftengine/src/bcstatetransfer/Messages.hpp

```diff
@@ -40,56 +40,55 @@ class MsgType {
 };
 
 struct BCStateTranBaseMsg {
+  BCStateTranBaseMsg(uint16_t type) : type(type) {}
   uint16_t type;
+  // This struct and its derived structs are used to de/serialize message buffers sent over communication channels.
+  // Since virtual methods modify the memory layout of a struct, there cannot be any for both this base struct and its
+  // inheritors.
 };
 
-struct AskForCheckpointSummariesMsg : public BCStateTranBaseMsg {
-  AskForCheckpointSummariesMsg() {
-    memset(this, 0, sizeof(AskForCheckpointSummariesMsg));
-    type = MsgType::AskForCheckpointSummaries;
+// MsgAllocator is useful for structs with a flexible array member. T should have a method setDataSize.
+template <typename T>
+class MsgAllocator {
+ public:
+  MsgAllocator(size_t dataSize) : memory_(new concord::Byte[calcBufferSize(dataSize)]()) {
+    (*this).operator->()->setDataSize(dataSize);
   }
 
-  uint64_t msgSeqNum;
-  uint64_t minRelevantCheckpointNum;
+  T* operator->() const { return reinterpret_cast<T*>(memory_.get()); }
+
+  concord::Byte* getBytes() const { return memory_.get(); }
```

Comment:
If we always cast to char* (I believe we do), I see no reason to have here concord::Byte. I suggest o hold
`std::unique_ptr<concord::char[]> buffer_;`
or
`std::unique_ptr<concord::char[]> data_;`
The rigght type would have been std::byte by the way, but if we can't change things, why not flow with what we have?